python/nav/ipdevpoll/plugins/paloaltoarp.py

```diff
@@ -35,75 +33,76 @@
-        )
         self._logger.debug("Collecting IP/MAC mappings for Paloalto device")
 
-        mappings = yield self._get_paloalto_arp_mappings(self.netbox.ip, api_key)
-        if mappings is None:
-            self._logger.info("No mappings found for Paloalto device")
-            returnValue(None)
+        configurations = yield self._get_paloalto_configurations(self.netbox)
+        for configuration in configurations:
+            mappings = yield self._get_paloalto_arp_mappings(
+                self.netbox.ip, configuration["api_key"]
+            )
+            if mappings:
+                yield self._process_data(mappings)
+                break
 
-        yield self._process_data(mappings)
+    @classmethod
+    @db.synchronous_db_access
+    def _has_paloalto_configurations(cls, netbox: Netbox):
+        """
+        Make a database request to check if the netbox has any management
+        profile that configures access to Palo Alto ARP data via HTTP
+        """
+        queryset = _paloalto_profile_queryset(netbox)
+        return queryset.exists()
 
-        returnValue(None)
+    @classmethod
+    @db.synchronous_db_access
+    def _get_paloalto_configurations(cls, netbox: Netbox):
+        """
+        Make a database request that fetches all management profiles of
+        the netbox that configures access to Palo Alto ARP data via HTTP
+        """
+        queryset = _paloalto_profile_queryset(netbox)
+        return list(queryset)
 
+    @classmethod
     @defer.inlineCallbacks
-    def _get_paloalto_arp_mappings(self, address: str, key: str):
-        """Get mappings from Paloalto device"""
-
-        arptable = yield self._do_request(address, key)
-        if arptable is None:
-            returnValue(None)
-
-        # process arpdata into an array of mappings
-        mappings = parse_arp(arptable.decode('utf-8'))
+    def _get_paloalto_arp_mappings(cls, address: IP, api_key: str):
+        """
+        Make a HTTP request for ARP data from Paloalto device with the given
+        ip-address, using the given api-key. Returns a formatted list of ARP
+        mappings for use in NAV.
+        """
+        arptable = yield cls._do_request(address, api_key)
+        mappings = _parse_arp(arptable) if arptable else []
         returnValue(mappings)
 
+    @classmethod
     @defer.inlineCallbacks
-    def _do_request(self, address: str, key: str):
-        """Make request to Paloalto device"""
+    def _do_request(cls, address: IP, key: str):
+        """Make a HTTP request to Paloalto device"""
 
         class SslPolicy(client.BrowserLikePolicyForHTTPS):
             def creatorForNetloc(self, hostname, port):
                 return ssl.CertificateOptions(verify=False)
 
         url = f"https://{address}/api/?type=op&cmd=<show><arp><entry+name+=+'all'/></arp></show>&key={key}"
-        self._logger.debug("making request: %s", url)
+        cls._logger.debug("making request: %s", url)
```

Comment:
This has an unfortunate side-effect: The logging statement will now *not* be in the context of the device being collected for, since you have stepped out of the class instance and into the class.

Maybe a linting tool complains that this can be a class method since it accesses no instance attributes, but I think maybe it should change back to keep the proper context.

Reply:
Also, we need to make a separate issue for something entirely unrelated to this PR: This logs the full contents of the *secret API token*, which is generally not a good idea from a security standpoint.  It should be censored.  Would you do the honors?

Reply:
Nice catch, reverted methods that uses logging back to being instance methods. The original intent was to avoid having to instantiate plugins during unit tests, which really is a non-issue anyways.

https://github.com/Uninett/nav/blob/c930c2f29efd399cf1a227a8246281d45b1534ba/python/nav/ipdevpoll/plugins/paloaltoarp.py#L93-L95

https://github.com/Uninett/nav/blob/c930c2f29efd399cf1a227a8246281d45b1534ba/tests/unittests/ipdevpoll/plugins_paloaltoarp_test.py#L121-L132